File: LeetCode/18-4sum/Solution.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * This class provides a solution to the kSum problem, specifically solving the 4Sum variant.
 */
public class Solution {

  /**
   * Finds all unique quadruplets in the array which sum up to the target.
   *
   * @param nums   - The input array of integers.
   * @param target - The target sum for the quadruplets.
   * @return A list of lists, where each list contains four integers that add up to the target.
   */
  public List<List<Integer>> fourSum(int[] nums, int target) {
    Arrays.sort(nums);
    return kSum(nums, (long) target, 4, 0);
  }

  /**
   * Finds all unique k-tuples in the array which sum up to the target.
   *
   * @param nums   - The input array of integers.
   * @param target - The target sum for the k-tuples.
   * @param k      - The number of elements in each tuple.
   * @param start  - The starting index in the array for the current kSum recursion.
   * @return A list of lists, where each list contains k integers that add up to the target.
   */
  private List<List<Integer>> kSum(int[] nums, long target, int k, int start) {
    List<List<Integer>> res = new ArrayList<>();
    if (start == nums.length || nums[start] * k > target || nums[nums.length - 1] * k < target) {
      return res;
    }

    if (k == 2) {
      return twoSum(nums, target, start);
    }

    for (int i = start; i < nums.length; i++) {
      if (i > start && nums[i] == nums[i - 1]) continue;
      for (List<Integer> subset : kSum(nums, target - nums[i], k - 1, i + 1)) {
        res.add(new ArrayList<>(Arrays.asList(nums[i])));
        res.get(res.size() - 1).addAll(subset);
      }
    }

    return res;
  }

  /**
   * Finds all unique pairs in the array which sum up to the target.
   *
   * @param nums   - The input array of integers.
   * @param target - The target sum for the pairs.
   * @param start  - The starting index in the array for the current twoSum calculation.
   * @return A list of lists, where each list contains two integers that add up to the target.
   */
  private List<List<Integer>> twoSum(int[] nums, long target, int start) {
    List<List<Integer>> res = new ArrayList<>();
    int lo = start, hi = nums.length - 1;

    while (lo < hi) {
      long sum = (long) nums[lo] + nums[hi];
      if (sum == target) {
        res.add(Arrays.asList(nums[lo], nums[hi]));
        while (lo < hi && nums[lo] == nums[lo + 1]) lo++;
        while (lo < hi && nums[hi] == nums[hi - 1]) hi--;
        lo++;
        hi--;
      } else if (sum < target) {
        lo++;
      } else {
        hi--;
      }
    }

    return res;
  }
}

File: LeetCode/18-4sum/solution.ts
function fourSum(nums: number[], target: number): number[][] {
  nums.sort((a, b) => a - b);
  return kSum(nums, target, 0, 4);
}

function kSum(
  nums: number[],
  target: number,
  start: number,
  k: number
): number[][] {
  let res: number[][] = [];
  // If we have run out of numbers to add, return res.
  if (start === nums.length) {
    return res;
  }
  // There are k remaining values to add to the sum. The
  // average of these values is at least target / k.
  let average_value = target / k;
  // We cannot obtain a sum of target if the smallest value
  // in nums is greater than target / k or if the largest
  // value in nums is smaller than target / k.
  if (nums[start] > average_value || average_value > nums[nums.length - 1]) {
    return res;
  }
  if (k === 2) {
    return twoSum(nums, target, start);
  }
  for (let i = start; i < nums.length; i++) {
    if (i === start || nums[i - 1] !== nums[i]) {
      for (let subset of kSum(nums, target - nums[i], i + 1, k - 1)) {
        res.push([nums[i], ...subset]);
      }
    }
  }
  return res;
}

function twoSum(nums: number[], target: number, start: number): number[][] {
  let res: number[][] = [];
  let lo: number = start,
    hi: number = nums.length - 1;
  while (lo < hi) {
    let curr_sum: number = nums[lo] + nums[hi];
    if (curr_sum < target || (lo > start && nums[lo] === nums[lo - 1])) {
      ++lo;
    } else if (
      curr_sum > target ||
      (hi < nums.length - 1 && nums[hi] === nums[hi + 1])
    ) {
      --hi;
    } else {
      res.push([nums[lo++], nums[hi--]]);
    }
  }
  return res;
}
